[BBQ] Extract Register Allocator for Shared Use

kmiller68 opened the pull request:
#### a0315f345b4e886dbfe4ebe74290ba315f3e1a1c
<pre>
[BBQ] Extract Register Allocator for Shared Use
<a href="https://bugs.webkit.org/show_bug.cgi?id=296341">https://bugs.webkit.org/show_bug.cgi?id=296341</a>
<a href="https://rdar.apple.com/156433214">rdar://156433214</a>

Reviewed by Dan Hecht.

This patch extracts BBQ&apos;s register allocator in to a shared header. There shouldn&apos;t be a significant
behavior change between the shared one and the new one. The new allocator takes a RegisterBank template
argument that defines the different details for both the JIT backend and register file in question.

The new class mostly follows the same terminology as BBQ did. The biggest change is that what was
called the LRU is now the Spiller, which operates on a spill hint. When we need to find a register
to evict we search for the value with the smallest hint. The highest bit is reserved for locking
a register or indicating that a register is ineligible. Additionally, the allocator has a abstract
RegisterBinding that it tracks and provides to the backend when requested or flushing. This binding
is intended to track any details the backend needs about the status of the register.

Lastly fix EnumTraits to work with Clang&apos;s safe buffer access for the enum&apos;s type name.

Canonical link: <a href="https://commits.webkit.org/297809@main">https://commits.webkit.org/297809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a01db469299139c6f70566cd5f5fee3064e1bcb9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63459 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33310 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41221 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85943 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115870 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101580 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66250 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25874 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19712 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62885 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105440 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95980 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19786 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122348 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111539 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40001 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/29827 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94801 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94539 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24138 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39665 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17485 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36114 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39887 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45386 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135769 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39528 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/36467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42861 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41265 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/jit/SimpleRegisterAllocator.h:
```diff
@@ -0,0 +1,267 @@
+        void setHint(Register reg, SpillHint newHint)
+        {
+            ASSERT(newHint >= 0);
+            ASSERT_IMPLIES(m_hints[reg] < 0, m_lockCounts[reg]); // Leave untracked registers alone.
```
what does untracked register mean in this context? Is it a synonym for locked?
And why is this an implication rather than equivalence like the others: ASSERT(!!m_lockCounts[reg] == (m_hints[reg] < 0)); Oh I guess maybe this is related to my question below.

dhecht commented on Source/JavaScriptCore/jit/SimpleRegisterAllocator.h:
```diff
@@ -0,0 +1,267 @@
+            m_hints[reg] &= ~lockBit;
+#if ASSERT_ENABLED
+            ASSERT(m_lockCounts[reg]);
+            m_lockCounts[reg]--;
```
so what exactly is the rule on pairing lock() and unlock() calls? That you can do a sequence of lock() calls, and then you have to do the same number of unlock() calls before another lock()? And that the register becomes unlocked on the first call to unlock(), not the last?

kmiller68 replied:
The intention is that you lock anything you need up front then do your bytecode/IR node then unlock everything you locked. Then you would move to the next opcode. The counting is just to ensure the lock and unlock calls match.

dhecht replied:
Right, but the confusion is around the side effect of unlock occurring on the first rather than last of these nested unlocks. The new comment clarifies the rule here (you can't allocate between the unlock calls).

dhecht commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5002,25 +4980,23 @@ Location BBQJIT::bind(Value value, Location loc)
     // for it here.
     ASSERT(!currentLocation.isRegister());
 
+    // The allocator could already have bound us to the
```
incomplete comment

dhecht reviewed (approved):
r=me

kmiller68 committed:
[BBQ] Extract Register Allocator for Shared Use
https://bugs.webkit.org/show_bug.cgi?id=296341
rdar://156433214

Reviewed by Dan Hecht.

This patch extracts BBQ's register allocator in to a shared header. There shouldn't be a significant
behavior change between the shared one and the new one. The new allocator takes a RegisterBank template
argument that defines the different details for both the JIT backend and register file in question.

The new class mostly follows the same terminology as BBQ did. The biggest change is that what was
called the LRU is now the Spiller, which operates on a spill hint. When we need to find a register
to evict we search for the value with the smallest hint. The highest bit is reserved for locking
a register or indicating that a register is ineligible. Additionally, the allocator has a abstract
RegisterBinding that it tracks and provides to the backend when requested or flushing. This binding
is intended to track any details the backend needs about the status of the register.

Lastly fix EnumTraits to work with Clang's safe buffer access for the enum's type name.

Canonical link: https://commits.webkit.org/297809@main
